chainer/function_node.py

```diff
@@ -1122,25 +1123,27 @@ def _backprop(outputs, inputs, grad_required, retain_grad, grads, loss_scale):
         # Do backward
 
         # Call pre-backward hooks
-        hooks = chainer.get_function_hooks()
         if func._n_local_function_hooks != 0:
-            hooks = collections.OrderedDict(hooks)
-            hooks.update(func.local_function_hooks)
-        hooks = hooks.values()  # avoid six for performance
+            local_hooks = collections.OrderedDict(chainer.get_function_hooks())
+            local_hooks.update(func.local_function_hooks)
+            hooks = local_hooks.values()  # avoid six for performance
+        else:
+            hooks = base_hooks
 
-        in_data = tuple([x.data for x in func.inputs])
-        out_grad_data = tuple(
-            [None if g is None else g.data for g in gys])
+        in_data = [x.data for x in func.inputs]
```

Comment:
Is it OK not to use `array` attribute instead of `data`?

Reply:
`x` is an instance of `VariableNode`, which only has `data`.

Reply:
Oops, got it.